backend/core/src/secondary-adapters/postgres/history/history-repository-adapter.ts

```diff
@@ -11,19 +13,81 @@ export class HistoryRepositoryAdapter implements HistoryRepository {
     this.prismaClient = prismaClient;
   }
 
-  getAllUserHistory(userId: string): Promise<History[]> {
-    return this.prismaClient.history.findMany({
+  async getAllUserHistoryLength(userId: string): Promise<number> {
+    return this.prismaClient.history.count({
       where: {
         userId,
       },
     });
   }
 
-  async createHistoryItem(historyRequestDto: HistoryRequestDto): Promise<HistoryResponseDto> {
-    const history = await this.prismaClient.history.create({
-      data: { ...historyRequestDto },
+  async getUserHistory(userId: string, take: number, skip: number, lastPage: number): Promise<HistoryResponseDto> {
+    const history = await this.prismaClient.history.findMany({
+      take,
+      skip,
+      where: {
+        userId,
+      },
+      include: {
+        video: {
+          include: {
+            channel: {
+              select: {
+                id: true,
+                name: true,
+                avatar: true,
+              },
+            },
+          },
+        },
+      },
+      orderBy: {
+        updatedAt: 'desc',
+      },
     });
+    const list: HistoryListType[] = history.map((historyRecord) => trimHistory(historyRecord));
 
-    return history;
+    return {
+      list,
+      lastPage,
+      currentPage: skip / 10 + 1,
+    };
+  }
+
+  isHistoryAlreadyExist(historyRequestDto: HistoryRequestDto): Promise<History | null> {
+    return this.prismaClient.history.findFirst({
+      where: {
+        userId: historyRequestDto.userId,
+        videoId: historyRequestDto.videoId,
+      },
+    });
+  }
+
+  updateHistoryRecord(id: string): Promise<History> {
+    return this.prismaClient.history.update({
+      where: {
+        id,
+      },
+      data: { updatedAt: new Date() },
+    });
```

Comment:
I've noticed you first check if the record exists in a query to db. then make another query to db to update/create record, depending on the result of the first query. Have you considered using `upsert`, which combines both operations? I'm not telling to rewrite it with `upsert`: if you intended to write it the way it is written now, just mark as a resolved.

Reply:
i tryed to use it but for upsert we need provide unique history fields userId and videoId are not unique according to the prisma scheme, there is only one unique field, this is id, but in this case it is necessary to update it by userId and videoId